### ISSUE TITLE
Wrong coverage metrics for FastAPI project

### ISSUE THREAD

**Issue by cashaev** [tiangolo/fastapi#3557]
Hi everyone. I am having an issue with pytest coverage report for my FastAPI project.
Part of the code, which is definitely executed is not market as covered. Namely return statements from routes.

Did anybody here have this issue? I have seen only one mention of it in stackoverflow, but not resolved.
Thank you

**Comment by Kludex** [comment]
Maybe related to https://github.com/nedbat/coveragepy/issues/1082 ?

**Comment by cashaev** [comment]
> Maybe related to [nedbat/coveragepy#1082](https://github.com/nedbat/coveragepy/issues/1082) ?

@Kludex Yes, looks like it. It was exactly my case: I have a FastAPI together with SQLAlchemy async calls. After reading the thread and other related threads carefully I noticed that coverage report in my project stopped after async calls to SQLAlchemy engine. As I understood the reason for it is that FastAPI uses threads, and SQLAlchemy uses greenlet to provide async behavior and coverage has problems tracking both in the same project.

I tried to play around with the coveragerc settings and came out using "concurrency=greenlet" config. So far, the report looks much better and shows hits on the code lines after SQLAlchemy async calls. I will check it further on.

Thank you for the link. I would definitely not have dug it out by myself!

**Comment by dgilmanTORQATA** [comment]
I'm also testing against a FastAPI project with a mix of async routes, async SQLAlchemy and synchronous FastAPI dependencies. Instructing coverage to use `concurrency = gevent` works in async routes that use SQLAlchemy but breaks coverage elsewhere in places like the FastAPI dependencies.

Adding a voodoo `await asyncio.sleep(0)` right after the last line that coverage picks up on (so right after the line that awaits on a SQLAlchemy coroutine) lets coverage get back up to speed and correctly detect coverage on the following lines.

**Comment by stephan-erb-by** [comment]
A fix for this has been merged to the coverage repository. Once the new coverage version is out you will be able to specify multiple `concurrency` options and thus have measurements work out of the box https://github.com/nedbat/coveragepy/commit/c9d821deba6f7ee5eef30fef5355f7c93808b4f9

**Comment by erewok** [comment]
After reading the above and making sure I had a recent version of coverage (`6.3.1` in my case), I was able to fix this on my project by adding the following to my `setup.cfg`:

```ini
[coverage:run]
concurrency =
  greenlet
  thread
```

**Comment by FilipeMarch** [comment]
How do you guys use coverage with FastAPI?
I am executing this on a terminal:
```shell
uvicorn app.main:app
```
And this on another terminal:
```
pytest -s --cov=. 
```
I get these results:
![image](https://user-images.githubusercontent.com/23220309/205241575-e579c67b-8ecd-41b3-8c1a-ca67d7a0e8cf.png)


But coverage does not show anything about my routes (`router.py` files):
![image](https://user-images.githubusercontent.com/23220309/205241840-e2b05029-5780-4de9-aba0-e38fb0bff56b.png)

Some files are showing 0% even though their lines are necessarily being executed (for instance, `main.py` is where the app starts). I would appreciate some teaching on how to use coverage correctly with FastAPI.
